### ISSUE TITLE
[feature request] Add YellowFin optimizer

### ISSUE THREAD

**Issue by ethancaballero** [pytorch/pytorch#1960]
from "YellowFin and the Art of Momentum Tuning" https://arxiv.org/abs/1706.03471

reference implementation:
https://github.com/JianGoForIt/YellowFin_Pytorch/blob/master/tuner_utils/yellowfin.py

cc: @JianGoForIt

**Comment by ethancaballero** [comment]
It would be convenient also if there was a built in way to share optimizer parameters (like momentum) among different threads like in A3C.

**Comment by jtoy** [comment]
is this something that the team would like to get added to pytorch?

**Comment by soumith** [comment]
Sure, we'd love to have it in, if anyone sends a PR. If not we'll eventually get to it.

**Comment by jonmorton** [comment]
The authors have now published a pytorch implementation of YellowFin at https://github.com/JianGoForIt/YellowFin_Pytorch

Maybe it just needs to be cleaned up a bit and reviewed to be integrated.

**Comment by bhushan23** [comment]
I am working on it.

**Comment by atulkum** [comment]
We should read this review comments about the paper before implementing it. It might not be worth implementing.
https://github.com/pytorch/pytorch/issues/1960

**Comment by EshwarSR** [comment]
Hi, I'm new to pytorch and would like to start contributing. If no one is working on it, I would like to work on this. Can one of you help me get started?

**Comment by bhushan23** [comment]
@EshwarSR you can take this task
New optimizers are placed here https://github.com/pytorch/contrib
You can go over existing optimizers https://github.com/pytorch/pytorch/tree/master/torch/optim and understand how they are implemented

**Comment by atulkum** [comment]
This is a bogus algorithm. https://openreview.net/forum?id=SyrGJYlRZ
As a community we should discourage these kind of `Misleading and shaky theoretical motivation/approach` even if it come from stanford ;)

**Comment by moon05** [comment]
@bhushan23 hey I am new to PyTorch and was also looking at this. I was wondering what needs to be different from the reference implementation here? It seems like other than the functions state_dict and load_state_dict, which is already done in optimizer.py everything else should stay the same. Or am I not understanding something here? Could you kindly clarify?

**Comment by kamathhrishi** [comment]
Could I go for this?

**Comment by bhushan23** [comment]
@kamathhrishi go for it

**Comment by kamathhrishi** [comment]
Alright then

**Comment by soumith** [comment]
@kamathhrishi i believe there are open PRs for this.

**Comment by soumith** [comment]
my bad, there aren't.

**Comment by kamathhrishi** [comment]
I will go for it , do you encourage draft PR's and where could we discuss specific implementation details ? Slack , PyTorch Forum or here?

**Comment by soumith** [comment]
draft PR seems fine.

**Comment by kamathhrishi** [comment]
I have begin some work on it and will take a while before I complete it.  Just to confirm YellowFin will be under torch.optim using SGD internally or under torch.optim.lr_scheduler ?

**Comment by soumith** [comment]
yellowfin is an optimizer, so it'll be a new file called yellowfin.py under optim. You need to read the paper and faithfully reproduce the algorithm as-is.

**Comment by kamathhrishi** [comment]
Alright then. The thing was that the paper explains YellowFin as a tuner which helps find optimal momentum and learning rate for momentum-based SGD.

**Comment by atulkum** [comment]
I think you should read this review before spending time on it.
https://openreview.net/forum?id=SyrGJYlRZ

**Comment by soumith** [comment]
the reviews look brutal. I've also checked on the general citation count and it's not very high.
@kamathhrishi how far are you into implementing it? I'm wondering if we shouldn't implement it into core.

**Comment by kamathhrishi** [comment]
@soumith Not much, I guess I will move onto working on some other PR.

**Comment by kamathhrishi** [comment]
@soumith Not much, how do you think I should go about it?

**Comment by soumith** [comment]
okay, let's not put YellowFin in core. Those who want it can still use the out of source implementations.

Since the issue opened (in June 2017), for newer feature requests (I think late 2017), we have been accepting them only if they become somewhat of a standard in the machine learning community (for example we didn't accept Swish and some other layers). So this decision is in line with those.

**Comment by kamathhrishi** [comment]
Alright
